tensorflow/core/kernels/data/dataset_test_base.h

```diff
@@ -0,0 +1,191 @@
+
+ protected:
+  // Creates a thread pool for parallel tasks.
+  Status InitThreadPool(int thread_num);
+
+  // Initializes the runtime for computing the dataset operation and registers
+  // the input function definitions. `InitThreadPool()' needs to be called
+  // before this method if we want to run the tasks in parallel.
+  Status InitFunctionLibraryRuntime(const std::vector<FunctionDef>& flib,
+                                    int cpu_num);
+
+  // Runs an operation producing outputs. The `context` need to be initialized
+  // (see `CreateOpKernelContext()` ) before running this method.
+  Status RunOpKernel(OpKernel* op_kernel, OpKernelContext* context);
+
+  // Checks that the size of `inputs` matches the requirement of the operation.
+  Status CheckOpKernelInput(const OpKernel& kernel,
+                            const gtl::InlinedVector<TensorValue, 4>& inputs);
+
+  // Creates a new context for running the dataset operation.
+  std::unique_ptr<OpKernelContext> CreateOpKernelContext(
+      OpKernel* kernel, gtl::InlinedVector<TensorValue, 4>* inputs);
+
+  // Returns a new serialization context for serializing the dataset and
+  // iterator.
+  std::unique_ptr<SerializationContext> CreateSerializationContext();
+
+  // Adds an arrayslice of data into the input vector. `input_types` describes
+  // the required data type for each input tensor. `shape` and `data` describes
+  // the shape and values of the current input tensor.
+  template <typename T>
+  void AddDatasetInputFromArray(gtl::InlinedVector<TensorValue, 4>* inputs,
+                                DataTypeVector input_types,
+                                const TensorShape& shape,
+                                const gtl::ArraySlice<T>& data) {
+    Tensor* input =
+        AddDatasetInput(inputs, input_types, DataTypeToEnum<T>::v(), shape);
+    test::FillValues<T>(input, data);
+  }
+
+ private:
+  // Adds an empty tensor to the input vector. The returned pointer can be used
+  // to fill in the value for the tensor.
+  Tensor* AddDatasetInput(gtl::InlinedVector<TensorValue, 4>* inputs,
+                          DataTypeVector input_types, DataType dtype,
+                          const TensorShape& shape);
+
+  // Sets the allocator attributes for the operation outputs.
+  void SetOutputAttrs();
+
+ protected:
+  std::unique_ptr<Device> device_;
+  DeviceType device_type_;
+  Allocator* allocator_;
+  std::vector<AllocatorAttributes> allocator_attrs_;
+  std::unique_ptr<ScopedStepContainer> step_container_;
+
+  FunctionLibraryRuntime* flr_;  // not owned
+  std::function<void(std::function<void()>)> runner_;
+  std::unique_ptr<DeviceMgr> device_mgr_;
+  std::unique_ptr<FunctionLibraryDefinition> lib_def_;
+  std::unique_ptr<OpKernelContext::Params> params_;
+  std::unique_ptr<ProcessFunctionLibraryRuntime> pflr_;
+  std::unique_ptr<thread::ThreadPool> thread_pool_;
+  std::vector<Tensor*> tensors_;  // owns Tensors.
```

Comment:
IIUC, this field only exists to deallocated the Tensors ... why are the Tensors created by as dynamic objects instead of values (which would be deallocated when they go out of scope)?

Reply:
Yeah, `tensors_ ` is used to deallocate the tensors as below. We don't want these tensors to be released when they go out of scope, because they will be used for the inputs of the op kernel in the next step.
```C++
gtl::InlinedVector<TensorValue, 4> inputs_;
...
Tensor* input = new Tensor(allocator_, dtype, shape);
tensors_.push_back(input);
inputs_->push_back({nullptr, input});
...
params_->inputs = inputs_;
...
gtl::STLDeleteElements(&tensors_);
```